### PR TITLE
added check for bind on create to determine local volume driver

### PR DESCRIPTION
**Pull request by clintkitson**
Signed-off-by: Clinton Kitson clintonskitson@gmail.com

### CODE REVIEW

**Comment by clintonskitson** [comment]
This patch fixes the issue https://github.com/docker/docker/issues/15467. Ping @calavera

The fix is to check what binds are specified and to only specify the selected volume driver when a bind is specified. Otherwise the default "local" volume driver is used.

For any volume that is specified in the Dockerfile but not binded, the "local" volume driver is used regardless.

It also ensures that contents of a container FS only attempt to be copied under the "local" volume driver scenario. I believe it is assumed that attached container volumes have persistent data and thus should not be overwritten by an initialization process.

**Comment by calavera** [comment]
This needs tests. We have a constant for the `local` driver in the volumes package, please use it.

**Comment by clintonskitson** [comment]
Ok, thanks for reviewing.  Will do.

**Comment by clintonskitson** [comment]
Question for you here @calavera.  As far as testing this, would you assume that integration-cli testing would suffice?  If so, it seems like creating a Dockerfile on the fly with a volume with data and then making a container from it to be tested with would be appropriate to test with?

**Comment by calavera** [comment]
yes, that sounds like a good plan. We already have tests that generate images on the fly in `integration-cli`, like in the [tests for docker rm](https://github.com/docker/docker/blob/master/integration-cli/docker_cli_rm_test.go#L43), but there are others. Feel free to copy the behavior in one of them.

**Comment by clintonskitson** [comment]
@calavera this works, but still needs some cleanup.  Passing it here to see if the method seems right.

Struggled through a bit to be able to get a container with a file at the right daemon.  I think creating this via an echo command in the Dockerfile seems legitimate.  The buildImage functions could use the args params to allow passing of the correct daemon socket.

**Comment by clintonskitson** [comment]
The error condition is
```
[d92493071] waiting for daemon to start
[d92493071] daemon started
docker_cli_start_volume_driver_unix_test.go:278:
    c.Assert(s.ec.creations, check.Equals, 1)
... obtained int = 2
... expected int = 1
```

File: integration-cli/docker_cli_start_volume_driver_unix_test.go [hunk 1]
```diff
@@ -244,3 +244,39 @@ func (s DockerExternalVolumeSuite) TestStartExternalVolumeDriverDeleteContainer(
 func hostVolumePath(name string) string {
 	return fmt.Sprintf("/var/lib/docker/volumes/%s", name)
 }
+
+func (s *DockerExternalVolumeSuite) TestStartExternalNamedVolumeDriverCheckBindLocalVolume(c *check.C) {
+
```
Comment (calavera) [inline integration-cli/docker_cli_start_volume_driver_unix_test.go]: can you remove this empty line?

File: integration-cli/docker_utils.go [hunk 2]
```diff
@@ -1353,3 +1353,33 @@ func createTmpFile(c *check.C, content string) string {
 
 	return filename
 }
+
+func buildImageArgs(args []string, name, dockerfile string, useCache bool) (string, error) {
+	id, _, err := buildImageWithOutArgs(args, name, dockerfile, useCache)
+	return id, err
+}
+
+func buildImageWithOutArgs(args []string, name, dockerfile string, useCache bool) (string, string, error) {
```
Comment (calavera) [inline integration-cli/docker_utils.go]: there are two things that puzzle me about this function:

1. `Without` is a word, and doesn't require the capitalized o.
2. The function suffix is without args and the first parameter is called args.

:confounded:

File: daemon/create_unix.go [hunk 3]
```diff
@@ -42,7 +43,15 @@ func createContainerPlatformSpecificSettings(container *Container, config *runco
 			return fmt.Errorf("cannot mount volume over existing file, file exists %s", path)
 		}
 
-		v, err := createVolume(name, config.VolumeDriver)
+		// if a bind is not specified then default to the local volume driver
+		var volumeDriver string
```
Comment (calavera) [inline daemon/create_unix.go]: you can set `volumeDriver := volume.DefaultDriverName` by default and avoid the `else`.

File: integration-cli/docker_utils.go (hunk 2 above)
Comment (clintonskitson) [inline integration-cli/docker_utils.go]: WithOut referred to WithOutput following from `buildImageWithOut`, or with output.  One of the questions was whether the existing methods should be enhanced in a separate PR or these that include the `Args` should be included here and supplement to existing ones.

Comment (calavera) [inline integration-cli/docker_utils.go]: ohh, that makes more sense, carry on, sorry for the noise.

**Comment by clintonskitson** [comment]
Updated based on comments.

**Comment by clintonskitson** [comment]
@calavera Notable change here is to include the image config directly with volumes defined as to whether to flip the default volume driver.

File: daemon/create_unix.go [hunk 4]
```diff
@@ -42,7 +45,19 @@ func createContainerPlatformSpecificSettings(container *Container, config *runco
 			return fmt.Errorf("cannot mount volume over existing file, file exists %s", path)
 		}
 
-		v, err := createVolume(name, config.VolumeDriver)
+		volumeDriver := config.VolumeDriver
+		if destination != "" && img != nil {
+			if _, ok := img.ContainerConfig.Volumes[destination]; ok {
+				logrus.Errorf("container1: v %s", img.ContainerConfig.Volumes[destination])
```
Comment (calavera) [inline daemon/create_unix.go]: is this a debug statement? is it necessary? the message is very misleading. Can we either use `Debugf` and a better message, or remove the log all together?

(Same three lines below :smile_cat: )

**Comment by clintonskitson** [comment]
Removed debugging statements.

**Comment by calavera** [comment]
LGTM

**Comment by jfrazelle** [comment]
LGTM

**Comment by cpuguy83** [comment]
One problem with this is, the data was only being copied IF the volume was empty, so there was no persistent data anyway.

**Comment by thaJeztah** [comment]
(we were just discussing this)

Basically, that change makes it impossible to replace the built-in volume driver ('local') with an alternative, and keep the same behavior.
